session/session.go

```diff
@@ -1241,11 +1243,19 @@ func (s *session) ParseSQL(ctx context.Context, sql, charset, collation string)
 	}
 	defer trace.StartRegion(ctx, "ParseSQL").End()
 
-	p := s.parserPool.Get().(*parser.Parser)
-	defer s.parserPool.Put(p)
+	p := parserPool.Get().(*parser.Parser)
+	defer parserPool.Put(p)
 	p.SetSQLMode(s.sessionVars.SQLMode)
 	p.SetParserConfig(s.sessionVars.BuildParserConfig())
-	return p.Parse(sql, charset, collation)
+	tmp, warn, err := p.Parse(sql, charset, collation)
+	// The []ast.StmtNode is referenced by the parser, to reuse the parser, make a copy of the result.
+	if len(tmp) == 1 {
+		s.cache[0] = tmp[0]
+		return s.cache[:], warn, err
+	}
```

Comment:
@tiancaiamao Hello, may I ask, what is the effect of this "cache" variable, because I see the use of this "cache" variable only here, shall we delete the cache and this conditional statement, it seems that if the length of tmp is one, we also can copy this slice and return.

Reply:
It said:

> // The []ast.StmtNode is referenced by the parser, to reuse the parser, make a copy of the result.

We can't simply return `tmp`, because it's referenced by the parser object.
So I make a copy here, from `tmp`  to `s.cache`

Of course you can try to make the copy  operation explicit by:
1. alloc a new slice
2. copy the data to that slice
3. return the slice

```
	if len(tmp) == 1 {
        cache = make([]ast.StmtNode, 1)
		cache[0] = tmp[0]
		return cache[:], warn, err
	}
```

That would make the allocation of `[]ast.StmtNode` every time.
To avoid that, a `cache [1]ast.StmtNode`  field is stored in the `session` and get reused everytime, to avoid the allocation.

Reply:
In 99.9% cases, the `Parse` result has only one ast.StmtNode, so the `if conditional statement` object reuse works.
For other cases, it just make the allocation...

If you don't want this optimization, it's possible to remove the `if conditional statement` and the code still works.